nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,446 @@
+						},
+					},
+				},
+				Required: true,
+			},
+		},
+	}
+}
+
+type NsxClusterNode struct {
+	ID        string
+	IPAddress string
+	UserName  string
+	Password  string
+	Fqdn      string
+	Status    string
+}
+
+func getClusterNodesFromSchema(d *schema.ResourceData) []NsxClusterNode {
+	nodes := d.Get("node").([]interface{})
+	var clusterNodes []NsxClusterNode
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		id := data["id"].(string)
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		fqdn := data["fqdn"].(string)
+		status := data["status"].(string)
+		nodeObj := NsxClusterNode{
+			ID:        id,
+			IPAddress: ipAddress,
+			UserName:  userName,
+			Password:  password,
+			Fqdn:      fqdn,
+			Status:    status,
+		}
+		clusterNodes = append(clusterNodes, nodeObj)
+	}
+	return clusterNodes
+}
+
+func setNodesInSchema(d *schema.ResourceData, nodes []NsxClusterNode) error {
+	// Retrieve node credential from schema and set them in the nodeList element
+	// This is because the nodes are obtained using cluster client, and does not
+	// contain any information about credential
+	nodeIPToCredentialMap := getIPtoCredentialMap(d)
+
+	var nodeList []map[string]interface{}
+	for _, node := range nodes {
+		_, nodeInSchema := nodeIPToCredentialMap[node.IPAddress]
+		elem := make(map[string]interface{})
+		elem["id"] = node.ID
+		elem["ip_address"] = node.IPAddress
+		// If node is not in schema but in cluster node list,
+		// it means the node needs to be removed.
+		// In this case we don't need its credential because
+		// remove api does not require node credential
+		if nodeInSchema {
+			elem["username"] = nodeIPToCredentialMap[node.IPAddress][0]
+			elem["password"] = nodeIPToCredentialMap[node.IPAddress][1]
+		}
+		elem["fqdn"] = node.Fqdn
+		elem["status"] = node.Status
+		nodeList = append(nodeList, elem)
+	}
+	return d.Set("node", nodeList)
+}
+
+func getIPtoCredentialMap(d *schema.ResourceData) map[string][]string {
+	// returns a map, whose key is IP address of a node,
+	// value is a slice of [node_login_username, node_login_password]
+	nodes := d.Get("node").([]interface{})
+	var idToCredentialMap = map[string][]string{}
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		cred := []string{userName, password}
+		idToCredentialMap[ipAddress] = cred
+	}
+	return idToCredentialMap
+}
+
+func resourceNsxtManagerClusterCreate(d *schema.ResourceData, m interface{}) error {
+	// Call Joincluster function on nodes that are not in the cluster
+	nodes := getClusterNodesFromSchema(d)
+	if len(nodes) == 0 {
+		return fmt.Errorf("Please provide at least one node ip")
+	} else if len(nodes) == 1 {
+		log.Printf("[INFO] Only one node ip provided, skip joining nodes process")
```

Comment:
I think this is not correct any longer, since we're not including main node in the list anymore?